File: KOTLIN/src/main/java/by/bsuir/library/model/CacheManager.kt
package by.bsuir.library.model

import android.content.Context
import com.google.firebase.storage.StorageReference
import kotlinx.coroutines.async
import kotlinx.coroutines.coroutineScope
import kotlinx.coroutines.tasks.await
import java.io.File
import java.util.Calendar

class CacheManager(context: Context, source: StorageReference, nMinutesTimeout: Int) {
    private val _cacheUpdaterSource: StorageReference = source
    private val _cacheDirectory: File = context.cacheDir
    private val _nMinutesCacheTimeout: Int = nMinutesTimeout
    private val _inMemoryCache: MutableMap<String, ByteArray> = mutableMapOf()

    init {
        _cacheDirectory.listFiles()?.forEach {
            _inMemoryCache[it.path.split("/").last()] = byteArrayOf()
        }
    }

    private fun add(data: ByteArray, name: String){
        _inMemoryCache[name] = data
        val toSave = File("${_cacheDirectory.path}/$name")

        toSave.outputStream().write(data)
    }

    suspend fun get(name: String): ByteArray = coroutineScope{
        if (_inMemoryCache.containsKey(name)) {
            val file = File("${_cacheDirectory.path}/$name")

            if (_inMemoryCache[name]!!.isEmpty()) {
                _inMemoryCache[name] = file.inputStream().readBytes()
            }

            val now = Calendar.getInstance().time.time
            val modified = file.lastModified()

            if ((now - modified) / 60000 > _nMinutesCacheTimeout) {
                add(loadFromStorage(name), name)
            }
        }
        else
        {
            add(loadFromStorage(name), name)
        }

        return@coroutineScope _inMemoryCache[name]!!
    }

    private suspend fun loadFromStorage(name: String): ByteArray{
        return _cacheUpdaterSource.child(name).getBytes(2 * 1024 * 1024).await()
    }
}

File: FLUTTER/lib/Model/cache_manager.dart
import 'dart:async';
import 'dart:io';
import 'dart:typed_data';

import 'package:firebase_storage/firebase_storage.dart';
import 'package:path_provider/path_provider.dart';

class CacheManager {
  late Directory _cacheDirectory;
  late Reference _cacheUpdaterSource;
  late int _nMinutesCacheTimeout;
  late Map<String, Uint8List> _inMemoryCache;

  CacheManager({required Reference cacheUpdaterSource, required int nMinutesCacheTimeout}) {
    getApplicationDocumentsDirectory().then((value) async {
      _cacheDirectory = value;
      _inMemoryCache = {};
      
      await _cacheDirectory.list(followLinks: false).forEach((element) {
        _inMemoryCache[element.path.split("/").last] = Uint8List(0);
      });
    });
    _cacheUpdaterSource = cacheUpdaterSource;
    _nMinutesCacheTimeout = nMinutesCacheTimeout;
  }

  Future add({required Uint8List data, required String name}) async {
    _inMemoryCache[name] = data;
    File toSave = File("${_cacheDirectory.path}/$name");
    
    await toSave.writeAsBytes(data);
  }

  Future<Uint8List> get({required String name}) async {
    if (_inMemoryCache.containsKey(name)) {
      File file = File("${_cacheDirectory.path}/$name");
      
      if (_inMemoryCache[name]!.isEmpty){
        _inMemoryCache[name] = file.readAsBytesSync();
      }

      if (file.lastModifiedSync().difference(DateTime.now()) > Duration(minutes: _nMinutesCacheTimeout)){
        add(data: await _loadFromStorage(name: name), name: name);
      }

    }
    else 
    {
      add(data: await _loadFromStorage(name: name), name: name);
    }

    return _inMemoryCache[name]!;
  }

  FutureOr<Uint8List> _loadFromStorage({required String name}) {
    return _cacheUpdaterSource.child(name).getData().then((value) {
      if (value != null) {
        return value;    
      } else {
        return Uint8List(0);
      }
    });
  }
}